pkg/k8s/watchers/callbacks/service.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package callbacks
+
+import (
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// ServiceHandler is implemented by event handlers responding to K8s Service
+// events.
+type ServiceHandler interface {
+	OnAdd(*slim_corev1.Service)
+	OnUpdate(oldObj, newObj *slim_corev1.Service)
+	OnDelete(*slim_corev1.Service)
+}
+
+// NewService creates a new callbacks list for ServiceHandlers.
+func NewService() *ServiceList {
+	return &ServiceList{
+		cbs: make([]ServiceHandler, 0),
+	}
+}
+
+// Register registers a ServiceHandler's callbacks for reacting to Service
+// objects into the list.
+func (l *ServiceList) Register(s ServiceHandler) {
+	l.Lock()
+	l.cbs = append(l.cbs, s)
+	l.Unlock()
+}
+
+// Callbacks returns all the Service event handler callbacks from the list.
+func (l *ServiceList) Callbacks() []ServiceHandler {
+	c := make([]ServiceHandler, 0)
+	l.RLock()
+	copy(c, l.cbs)
```

Comment:
Nothing will be copied [if `len(l.cbs) == 0`](https://golang.org/pkg/builtin/#copy).
```suggestion
	l.RLock()
	c := make([]ServiceHandler, len(l.cbs))
	copy(c, l.cbs)
```

Reply:
Thanks! Yes I caught this earlier, and updated https://github.com/cilium/cilium/pull/15295, but didn't rebase this PR on top. I didn't want to push too much to this PR in case other were reviewing at the moment.